### PR TITLE
MSVC check for installed versions - VS2017 in particular

### PR DESCRIPTION
**Pull request by dmoody256**
mailing list: https://pairlist4.pair.net/pipermail/scons-users/2019-January/007513.html

Things added in this PR:

[**src**] changes:
- Update vc.py to check for cl.exe for VC 14.1 and newer, which use a new directory structure when searching for installed MSVC's.
- Updated vswhere call to include finding Visual Studio build tools
- Added 'arm' and 'arm64' target platforms as options to the msvc tool. Also updated vc.py to consider the platform when looking if a valid cl.exe is available when searching for installed MSVC's.
- Updated function comments in vc.py and msvc.xml tool documentation.
- Updated common.py to include the UWP environment variable when extracting vcvars script enivronment variables to the construction environment.
- updated find_vs_dir_by_reg from vs.py to also use the VC dir as a way to find installed Visual Studios.

[**test**] changes:
- added 4 tests to check Visual Studio 2017 project generation
- updated UWP test to work with Visual Studio 2017 which uses new environment variable
- updated TARGET_ARCH.py test to make sure arm and arm64 targets are working (only tested if installed)
- updated TestSconsMSVS.py to include new project templates.

need to:
- [x] check cl.exe is in detected VC dir
- [x] if a TARGET_ARCH is in the env, check if the corresponding cl.exe exists
- [x] Fix MSVC_UWP_APP.py test so libfile paths work for msvc 14.1
- [x] Add support for arm targets (arm and arm64)
- [x] Add msvc tests for MSVC 14.1 and add test to check TARGET_ARCH is checking the right toolset for msvc

### CODE REVIEW

File: src/engine/SCons/Tool/MSCommon/vc.py [hunk 1]
```diff
@@ -344,29 +347,99 @@ def find_batch_file(env,msvc_version,host_arch,target_arch):
...
+            host_dir = "Hostx86"
+        else:
+            return False
+
+        if target_platform in ('amd64','x86_64'):
+            target_dir = "x64"
+        elif target_platform in ('i386','i686','x86'):
+            target_dir = "x86"
+        else:
+            return False
```
Comment (bdbaddog) [inline src/engine/SCons/Tool/MSCommon/vc.py]: should we also have ARM here?

Comment (dmoody256) [reply src/engine/SCons/Tool/MSCommon/vc.py]: yes, still need to do that, I don't have access to a machine that has visual studio installed to help me do it atm

**Comment by dmoody256** [comment]
adding arm support would be related to https://github.com/SCons/scons/issues/3182

**Comment by dmoody256** [comment]
@bdbaddog could you close some of those sider issues? They are line to long issues in the VS project template tests.

**Comment by dmoody256** [comment]
should fix #3268

**Review by bdbaddog — COMMENTED** [review]
> @bdbaddog could you close some of those sider issues? They are line to long issues in the VS project template tests.

Done

File: src/engine/SCons/Tool/MSCommon/vc.py (hunk 1 above)
Comment (dmoody256) [reply src/engine/SCons/Tool/MSCommon/vc.py]: yeah they are version specific, so it would require two dicts, but its still makes the code more organized

**Comment by bdbaddog** [comment]
Can you summarize at the top what you changed?

**Comment by dmoody256** [comment]
> Can you summarize at the top what you changed?

done and updated documentation

File: src/engine/SCons/Tool/MSCommon/vc.py [hunk 2]
```diff
@@ -80,25 +80,76 @@ class BatchFileExecutionError(VisualCException):
...
     "x86_64"    : "amd64",
-    "x86_amd64" : "x86_amd64", # Cross compile to 64 bit from 32bits
+    "arm"       : "arm",
+    "arm64"     : "arm64",
+    "aarch64"   : "arm64",
+}
+ 
+# get path to the cl.exe dir for newer VS versions
+# based off a tuple of (host, target) platforms
+_HOST_TRGT_TO_CL_DIR_GREATER_THAN_14 = {
```
Comment (bdbaddog) [inline src/engine/SCons/Tool/MSCommon/vc.py]: change TRGT to TARGET please (also below)

File: src/engine/SCons/Tool/MSCommon/vc.py [hunk 3]
```diff
@@ -344,29 +413,132 @@ def find_batch_file(env,msvc_version,host_arch,target_arch):
...
+    ver_num = float(get_msvc_version_numeric(msvc_version))
+
+    # make sure the cl.exe exists meaning the tool is installed
+    if ver_num > 14:
+        # 2017 and newer allowed multiple versions of the VC toolset to be installed at the same time.
+        # Just get the default tool version for now
+        #TODO: support setting a specific minor VC version
+        default_toolset_file = os.path.join(vc_dir, r'Auxiliary\Build\Microsoft.VCToolsVersion.default.txt')
+        try:
+            f = open(default_toolset_file)
```
Comment (bdbaddog) [inline src/engine/SCons/Tool/MSCommon/vc.py]: Change to with open().. as f: context manager?

Comment (dmoody256) [reply src/engine/SCons/Tool/MSCommon/vc.py]: updated in f47c258

File: src/engine/SCons/Tool/MSCommon/vc.py (hunk 2 above)
Comment (dmoody256) [reply src/engine/SCons/Tool/MSCommon/vc.py]: updated in f47c258

**Comment by bdbaddog** [comment]
Great work! 
Merging!
